### PR TITLE
Add dead:disable comment

### PR DESCRIPTION
**Pull request by Afonasev**
Hello. I want use dead for my CI. Can i add nonzero return value and disable comment?

### CODE REVIEW

File: dead.py [hunk 1]
```diff
@@ -23,29 +24,38 @@
...
         self.filename = ''
+        self.lines: List[str] = []
         self.is_test = False
         self.reads: UsageMap = collections.defaultdict(set)
         self.defines: UsageMap = collections.defaultdict(set)
         self.reads_tests: UsageMap = collections.defaultdict(set)
 
+    def disabled(self, node: ast.AST) -> bool:
+        line = self.lines[node.lineno - 1]
+        return bool(DISABLE_COMMENT_RE.search(line))
```
Comment (asottile) [inline dead.py]: I think it should still identify all things that are unused, but at the end filter them out?  that way the information can still be retained

File: dead.py [hunk 2]
```diff
@@ -203,21 +213,30 @@ def main(argv: Optional[Sequence[str]] = None) -> int:
...
+
+        with visitor.file_ctx(filename, lines, is_test=is_test):
             visitor.visit(tree)
 
             with open(filename, 'rb') as f:
                 for tp, s, (lineno, _), _, _ in tokenize.tokenize(f.readline):
                     if tp == tokenize.COMMENT:
                         visitor.visit_comment(lineno, s)
 
+    has_errors = False
```
[the remark this answers is not in the document]
Comment (Afonasev) [reply dead.py]: i created separate PR https://github.com/asottile/dead/pull/12

**Comment by Afonasev** [comment]
I have updated PR, please look

File: dead.py [hunk 3]
```diff
@@ -200,28 +232,17 @@ def main(argv: Optional[Sequence[str]] = None) -> int:
...
-    for filename, is_test in _filenames(files_re, exclude_re, tests_re):
-        tree = _ast(filename)
 
-        with visitor.file_ctx(filename, is_test=is_test):
-            visitor.visit(tree)
-
-            with open(filename, 'rb') as f:
-                for tp, s, (lineno, _), _, _ in tokenize.tokenize(f.readline):
-                    if tp == tokenize.COMMENT:
-                        visitor.visit_comment(lineno, s)
```
Comment (asottile) [inline dead.py]: why did you change this? all you needed to do was add functionality to `visit_comment` :(

File: dead.py [hunk 4]
```diff
@@ -17,21 +17,26 @@
 from identify.identify import tags_from_path
 
 UsageMap = DefaultDict[str, Set[str]]
+DefineSet = Set[Tuple[str, str, int]]
+DisableSet = Set[Tuple[str, int]]
```
Comment (asottile) [inline dead.py]: there's no point in making a type alias if it's only used once

File: dead.py [hunk 5]
```diff
@@ -17,21 +17,26 @@
...
+
```
Comment (asottile) [inline dead.py]: ?

File: dead.py [hunk 6]
```diff
@@ -17,21 +17,26 @@
...
+DISABLE_COMMENT_RE = re.compile(r'dead\s*:\s*disable')
 
 
 class Visitor(ast.NodeVisitor):
     def __init__(self) -> None:
         self.filename = ''
         self.is_test = False
         self.reads: UsageMap = collections.defaultdict(set)
-        self.defines: UsageMap = collections.defaultdict(set)
+        self.defines: DefineSet = set()
```
Comment (asottile) [inline dead.py]: why did this change?

File: tests/visitor_test.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        path.write(src)
+        visitor.visit_file(path, is_test=is_test)
+
+    errors = list(find_unused(visitor))
+
+    if error:
+        assert errors
+        assert re.search(error, errors[0])
+    else:
+        assert not errors
```
Comment (asottile) [inline tests/visitor_test.py]: don't write logic in tests, this should be two separate tests

File: tests/visitor_test.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    for filename, (src, is_test) in files.items():
+        path = tmpdir.join(filename)
+        path.write(src)
+        visitor.visit_file(path, is_test=is_test)
+
+    errors = list(find_unused(visitor))
+
+    if error:
+        assert errors
+        assert re.search(error, errors[0])
```
Comment (asottile) [inline tests/visitor_test.py]: use exact equality

File: tests/visitor_test.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+import pytest
+
+from dead import find_unused
+from dead import Visitor
+
+
+@pytest.fixture()
+def visitor() -> Visitor:
+    return Visitor()
```
Comment (asottile) [inline tests/visitor_test.py]: this is not a useful fixture, you've added 5 lines to save 1 (+4)

File: dead.py (hunk 3 above)
Comment (Afonasev) [reply dead.py]: it is more convenient to write tests for me

File: dead.py (hunk 4 above)
Comment (Afonasev) [reply dead.py]: fixed

File: tests/visitor_test.py (hunk 7 above)
Comment (Afonasev) [reply tests/visitor_test.py]: done

File: dead.py (hunk 6 above)
Comment (Afonasev) [reply dead.py]: because values from that UsageMap not used

**Comment by asottile** [comment]
thanks!
